crypto/evp/evp_exchange.c

```diff
@@ -0,0 +1,354 @@
+{
+    int ref = 0;
+
+    CRYPTO_UP_REF(&exchange->refcnt, &ref, exchange->lock);
+    return 1;
+}
+
+EVP_EXCHANGE *EVP_EXCHANGE_fetch(OPENSSL_CTX *ctx, const char *algorithm,
+                                 const char *properties)
+{
+    EVP_EXCHANGE *exchange =
+        evp_generic_fetch(ctx, OSSL_OP_EXCHANGE, algorithm, properties,
+                          evp_exchange_from_dispatch,
+                          (int (*)(void *))EVP_EXCHANGE_up_ref,
+                          (void (*)(void *))EVP_EXCHANGE_free);
+
+
+    return exchange;
+}
+
+int EVP_PKEY_derive_init_ex(EVP_PKEY_CTX *ctx, EVP_EXCHANGE *exchange)
+{
+    int ret;
+    OSSL_PARAM *param = NULL;
+    size_t paramsz = 0;
+
+    ctx->operation = EVP_PKEY_OP_DERIVE;
+
+    if (ctx->engine != NULL)
+        goto legacy;
+
+    if (exchange != NULL) {
+        if (!EVP_EXCHANGE_up_ref(exchange))
+            goto err;
+    } else {
+        int nid = ctx->pkey != NULL ? ctx->pkey->type : ctx->pmeth->pkey_id;
+
+        /*
+         * TODO(3.0): Check for legacy handling. Remove this once all all
+         * algorithms are moved to providers.
+         */
+        if (ctx->pkey != NULL) {
+            switch (ctx->pkey->type) {
+            case EVP_PKEY_DH:
+                break;
+            default:
+                goto legacy;
+            }
+            /*
+             * If the EVP_PKEY has not been set, then this is a legacy use of
+             * the key exchange functions (scrypt/tls1_prf/hkdf). They're really
+             * KDFs. We don't want explicit fetches to find these "key exchange"
+             * algorithms, so we add the "internal" property.
+             */
+            exchange = EVP_EXCHANGE_fetch(NULL, OBJ_nid2sn(nid),
+                                          ctx->pkey == NULL ? "internal" : NULL);
+        } else {
+            goto legacy;
+        }
+
+        if (exchange == NULL) {
+            EVPerr(EVP_F_EVP_PKEY_DERIVE_INIT_EX, EVP_R_INITIALIZATION_ERROR);
+            goto err;
+        }
+    }
+
+    if (ctx->exchprovctx != NULL && ctx->exchange != NULL)
+        ctx->exchange->freectx(ctx->exchprovctx);
+    EVP_EXCHANGE_free(ctx->exchange);
+    ctx->exchange = exchange;
+    if (ctx->pkey != NULL) {
+        param = evp_pkey_to_param(ctx->pkey, &paramsz);
+        if (param == NULL) {
+            EVPerr(EVP_F_EVP_PKEY_DERIVE_INIT_EX, EVP_R_INITIALIZATION_ERROR);
+            goto err;
+        }
+    }
+    ctx->exchprovctx = exchange->newctx(ossl_provider_ctx(exchange->prov));
+    if (ctx->exchprovctx == NULL) {
+        OPENSSL_clear_free(param, paramsz);
+        EVPerr(EVP_F_EVP_PKEY_DERIVE_INIT_EX, EVP_R_INITIALIZATION_ERROR);
+        goto err;
+    }
+    ret = exchange->init(ctx->exchprovctx, param);
+    OPENSSL_free(param);
+
+    return ret ? 1 : 0;
+ err:
+    ctx->operation = EVP_PKEY_OP_UNDEFINED;
+    return 0;
+
+ legacy:
+    if (!ctx || !ctx->pmeth || !ctx->pmeth->derive) {
```

Comment:
may as well fix up the NULL checks here..

Reply:
Fixed.